File: Cplusplus/0086_partition_list/0086_partition_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        if (head == NULL || head->next == NULL)
            return head;
        ListNode* small = new ListNode(0);
        ListNode* ts = small;  
        ListNode* large = new ListNode(0);
        ListNode* tl = large;
        while (head != 0) {
            if (head->val < x) {
                ts->next = head;
                ts = ts->next;
            } else {
                tl->next = head;
                tl = tl->next;
            }
            head = head->next;
        }
        tl->next = NULL; 
        ts->next = large->next; 
        return small->next;
    }
};

File: Python/0086_partition_list/0086_partition_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        if head == None or head.next == None:  
            return head 
        small = ListNode(0) 
        ts = small  
        large = ListNode(0) 
        tl = large  
        while head != None: 
            if head.val < x:  
                ts.next = head 
                ts = ts.next  
            else:
                tl.next = head
                tl = tl.next 
            head = head.next    
        tl.next = None  
        ts.next = large.next 
        return small.next
